libraries/chain/controller.cpp

```diff
@@ -4066,6 +4066,15 @@ struct controller_impl {
                   "QC is_strong (${s1}) in block extension does not match is_strong_qc (${s2}) in header extension. Block number: ${b}",
                   ("s1", qc_proof.is_strong())("s2", new_qc_claim.is_strong_qc)("b", block_num) );
 
+      if (prev.valid) {
```

Comment:
I may have missed something here, but this doesn't look thread safe. This can be filled in when validating the block which could be happening at the same time this is running on the net thread. Can you instead check `validated` and then `assert(prev.valid)`? Is it correct that `prev.valid` is set when `validated==true`?

Reply:
`valid` https://github.com/AntelopeIO/spring/blob/e43fd1e468276e1791471159e2e75503f13db6aa/libraries/chain/include/eosio/chain/block_state.hpp#L77 is for Finality Merkle Tree while `validated`https://github.com/AntelopeIO/spring/blob/e43fd1e468276e1791471159e2e75503f13db6aa/libraries/chain/include/eosio/chain/block_state.hpp#L81 is mostly for forkbd. They are independent. Since I am using previous block's `valid` structure, it should be already constructed. I don't see a thread safe problem.

Reply:
During the transition to Savanna it is modified here:
https://github.com/AntelopeIO/spring/blob/53ad484eb68863d8aac86151e0c71c18de7fcf63/libraries/chain/controller.cpp#L1385-L1385
https://github.com/AntelopeIO/spring/blob/53ad484eb68863d8aac86151e0c71c18de7fcf63/libraries/chain/controller.cpp#L1406-L1406

Even outside Savanna transition it is modified here:
https://github.com/AntelopeIO/spring/blob/53ad484eb68863d8aac86151e0c71c18de7fcf63/libraries/chain/controller.cpp#L804-L804
This can happen while the net threads are accessing `prev.valid`.

I believe it is correct that `valid` will be filled in when `validated` is true. Since `validated` is atomic you could check that value here instead of `valid`.